CC: Software Sustainability Matrix (#1833) (#1844)
betterscientificsoftware/bssw.io: CuratedContent/SoftwareSustainabilityMatrix.md

<!--deck text end-->

#### Contributed by [Roscoe A. Bartlett](https://github.com/bartlettroscoe "Roscoe A. Bartlett GitHub Profile")
#### Publication date: ???

Resource information | Details
:--- | :--- 
Article title |  [How Sustainable is Your Software?](https://www.kitware.com/how-sustainable-is-your-software/)
Authors | Will Schroeder, Jess Tate, Jean-Christophe Fillion-Robin, Christopher Johnson, Dženan Zukić, Matt McCormick, Lee A. Newberg, Daniel White, Alexandra Warner, Ross Whitaker and Rob MacLeod
Publication Details | [Kitware Blog](https://www.kitware.com/blog), January 3, 2020
Focus | Software sustainability

The Software Sustainability Matrix (SSM) is a joint effort of the Scientific Computing and Imaging (SCI) Institute and Kitware to improve the health of the open-source package ecosystem making up the foundational software portfolio for the Center for Integrative Biomedical Computing (CIBC) lead by the SCI for the National Institutes of Health (NIH) <sup>[1],[2]</sup>.
The goal of the SSM effort is to create numerical scores for different aspects of software sustainability for a software package as well of a combined sustainability score.

> These SSM sustainability scores are being to guide efforts to improve the sustainability of the software package ecosystem making to CIBC software portfolio, with the hopes that it will encourage support and contributions from the open-source community and reduce the cost and risks to sustain this collection software.

Reviewer: @bartlettroscoe I tried to make a suggestion to fix this sentence starting with changing "being" to "being used", but then stumbled on what you are saying by "making to CIBC software porfolio". I think you will need to fix it.

Author: @pagrubel, I reworded this sentence in commit e325da88e801036e9a5daf42009a94239c9f5437.  Please take a look.

However, the SSM approach may be applicable to other package ecosystems and therefore may be of more general interest.

The SSM score is composed from component scores in the four areas Impact, Risks, Community, and Technology as shown in [Figure 1](#fig_software_sustainability_matrix) and each of these areas are broken down into several different subareas.

<a name="fig_software_sustainability_matrix"/>

<br>
<center>
<img src='../images/SoftwareSustainabilityMatrix.png' class='page'/>
</center>
<br>

<center>
<b>Figure 1:</b> Software Sustainability Matrix
</center>

<br>
<br>
 
Scores in the range [0-100] in these four areas Impact, Risks, Community, and Technology are then combined in a weighted sum:

&nbsp;&nbsp;&nbsp;&nbsp;SSM Score = F<sup>I</sup> * I + F<sup>R</sup> * R + F<sup>C</sup> * C + F<sup>T</sup> * T

to produce a single SSM Score [0-100], where F<sup>I</sup> + F<sup>R</sup> + F<sup>C</sup> + F<sup>T</sup> = 1.
The SCI & Kitware process<sup>[3]</sup> typically uses the weighting factors F<sup>I</sup> = F<sup>C</sup> = 1/3, and F<sup>R</sup> = F<sup>T</sup> = 1/6.

> It is not specified how the SCI & Ktiware program computes the component scores in the four areas 'I', 'R', 'C', and 'T' for the software packages of interest and the authors state that computing these individual component scores is ultimately a subjective matter.

Reviewer: ```suggestion
It is not specified how the SCI & Kitware program computes the component scores in the four areas 'I', 'R', 'C', and 'T' for the software packages of interest, and the authors state that computing these individual component scores is ultimately a subjective matter.
```

Author: I applied the suggested change in commit e325da88e801036e9a5daf42009a94239c9f5437.

It is stated by the authors that the primary benefit of estimating the SSM score by a software package team is not in computing the final score, but instead in highlighting areas of possible improvement.
It is stated that many teams make significant improvements in the sustainability for their packages while working through in the process of estimating the SSM score by adopting better processes in different areas.
In this respect, this is similar to the impact of the OpenSSF Best Practices Badge Program<sup>[4]</sup>.
That is, by having a team fill out each best practice item, the team will often implement suggested best practices, especially for practices that are not too difficult to adopt and support.

> In summary, the Software Sustainability Matrix (SSM) is a promising approach to help quantify the sustainability of individual software packages (and collectively entire software package ecosystems) and then help to drive efforts to make targeted improvements to improve sustainability.

Reviewer: ```suggestion
In summary, the Software Sustainability Matrix (SSM) is a promising approach to help quantify the sustainability of individual software packages (and collectively entire software package ecosystems) and subsequently help to drive efforts to make targeted improvements for sustainability.
```

Author: I applied the suggested change in commit e325da88e801036e9a5daf42009a94239c9f5437.

Publication Details | [Kitware Blog](https://www.kitware.com/blog), January 3, 2020
Focus | Software sustainability

The Software Sustainability Matrix (SSM) is a joint effort of the Scientific Computing and Imaging (SCI) Institute and Kitware to improve the sustainability of the open-source package ecosystem making up the foundational software portfolio for the Center for Integrative Biomedical Computing (CIBC) lead by the SCI for the National Institutes of Health (NIH) <sup>[1],[2]</sup>.
The goal of the SSM effort is to create numerical scores for different aspects of software sustainability for a software package as well of a combined sustainability score.

> This SSM process is being used to guide efforts to improve the sustainability of the CIBC software package ecosystem, with the hopes to transition the continuing support of this software to the open-source community<sup>[1]</sup>.

Author: Addressed in new commit new commit 071b872eebba57400c7233f0665be53bb7e2f6db.

to produce a single SSM Score [0-100], where F<sup>I</sup> + F<sup>R</sup> + F<sup>C</sup> + F<sup>T</sup> = 1.
The SCI & Kitware process<sup>[3]</sup> typically uses the weighting factors F<sup>I</sup> = F<sup>C</sup> = 1/3, and F<sup>R</sup> = F<sup>T</sup> = 1/6.
It is not specified how the SCI & Kitware program computes the component scores in the four areas 'I', 'R', 'C', and 'T' for the software packages of interest, and the authors state that computing these individual component scores is ultimately a subjective matter.

It is stated by the authors that the primary benefit of estimating the SSM score by a software package team is not in computing the final score, but instead in highlighting areas of possible improvement.

> It is claimed that many teams make significant improvements in the sustainability for their packages while working through in the process of estimating the SSM score by adopting better processes in different areas.

Author: Addressed in new commit new commit 071b872eebba57400c7233f0665be53bb7e2f6db.

In this respect, this is similar to the impact of the OpenSSF Best Practices Badge Program<sup>[4]</sup>.
That is, by having a team score each suggested practice, the team will often implement many suggested best practices, especially for practices that are not too difficult to adopt and support.

> In summary, the Software Sustainability Matrix (SSM) is a promising approach to help quantify the sustainability of individual software packages (and collectively entire software package ecosystems) and subsequently to drive efforts to make targeted improvements for sustainability.

Author: Addressed in new commit new commit 071b872eebba57400c7233f0665be53bb7e2f6db.
